juju/juju
# [JUJU-2741] Dqlite node config - detect if host is already a node

This is a preparatory patch, which does two main things:
- The `OptionFactory` is renamed to `NodeManager` in anticipation of its wider set of responsibilities.
- Said `NodeManager` includes a new method `IsExistingNode`, which is used by the db-accessor worker to eschew supplying options for bind address and cluster, if the host has previously been configured as a Dqlite node.

In addition, some of the shim infrastructure that we do not use has been removed from the `database` package in the interests of brevity.

## QA steps

- Bootstrap to LXD and enable HA.
- Once settled, run `juju debug-log -m controller --replay | grep dbaccessor`.
- You should observe the bootstrap node log the fact that it is already configured, followed by machines 1 and 2 indicating that they are being set up for the first time (abridged output):
```
machine-0: 12:45:05 INFO juju.worker.dbaccessor host is configured as a Dqlite node
machine-0: 12:45:05 INFO juju.worker.dbaccessor initialized Dqlite application (ID: 3297041220608546238)
...
machine-1: 12:46:36 INFO juju.worker.dbaccessor configuring host as a Dqlite node for the first time
machine-1: 12:46:36 DEBUG juju.worker.dbaccessor chose Dqlite bind address 10.246.27.109 from [10.246.27.109]
machine-1: 12:46:36 DEBUG juju.worker.dbaccessor determined Dqlite cluster members: [10.246.27.249:17666]
machine-1: 12:46:36 INFO juju.worker.dbaccessor initialized Dqlite application (ID: 17128632705061699021)
...
machine-2: 12:46:42 INFO juju.worker.dbaccessor configuring host as a Dqlite node for the first time
machine-2: 12:46:42 DEBUG juju.worker.dbaccessor chose Dqlite bind address 10.246.27.187 from [10.246.27.187]
machine-2: 12:46:42 DEBUG juju.worker.dbaccessor determined Dqlite cluster members: [10.246.27.249:17666]
machine-2: 12:47:27 INFO juju.worker.dbaccessor initialized Dqlite application (ID: 5758776007767032721)
```

## Documentation changes

None.

## Bug reference

N/A
